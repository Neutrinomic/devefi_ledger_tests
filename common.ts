import { Principal } from '@dfinity/principal';
import { resolve } from 'node:path';
import { Actor, PocketIc, createIdentity} from '@dfinity/pic';
import { IDL } from '@dfinity/candid';
import { _SERVICE as BasicService, idlFactory as BasicIdlFactory, init as basicInit } from './build/basic.idl.js';
import { _SERVICE as BurnService, idlFactory as BurnIdlFactory, init as burnInit } from './build/burn.idl.js';
import { _SERVICE as FastscanService, idlFactory as FastscanIdlFactory, init as fastscanInit } from './build/fastscan.idl.js';
import { _SERVICE as PassbackService, idlFactory as PassbackIdlFactory, init as passbackInit } from './build/passback.idl.js';
import { _SERVICE as ICRCLedgerService, idlFactory as ICRCLedgerIdlFactory, init as icrcInit, LedgerArg as ICRCLedgerArg, GetTransactionsRequest, GetTransactionsResponse, Transaction as ICRCTransaction, Account } from './icrc_ledger/ledger.idl.js';
import { _SERVICE as ICPLedgerService, idlFactory as ICPLedgerIdlFactory, init as icpInit, LedgerCanisterPayload as ICPLedgerCanisterPayload, QueryBlocksResponse, Block as ICPBlock, Transaction as ICPTransaction } from './icp_ledger/ledger.idl';
import { _SERVICE as CycleWalletService, idlFactory as CycleWalletIdlFactory, init as cycleWalletInit } from './build/cycle_wallet.idl.js';
import { _SERVICE as NTCService, idlFactory as NTCIdlFactory, init as ntcInit } from './build/NTC.idl.js';
import { _SERVICE as NTCCHAT_SERVICE, idlFactory as NTCCHAT_IDL_FACTORY, init as ntcchat_init } from './build/ntcchat.idl.js';

//@ts-ignore
import {toState} from "@infu/icblast";
import { AccountIdentifier, SubAccount } from '@dfinity/ledger-icp';
// Jest can't handle multi threaded BigInts o.O That's why we use toState

export const BASIC_WASM_PATH = resolve(__dirname, "./build/basic.wasm");
export const BURN_WASM_PATH = resolve(__dirname, "./build/burn.wasm");
export const FASTSCAN_WASM_PATH = resolve(__dirname, "./build/fastscan.wasm");
export const PASSBACK_WASM_PATH = resolve(__dirname, "./build/passback.wasm");
export const NTC_WASM_PATH = resolve(__dirname, "./build/NTC.wasm");
export const CYCLE_WALLET_WASM_PATH = resolve(__dirname, "./build/cycle_wallet.wasm");
export const NTCCHAT_WASM_PATH = resolve(__dirname, "./build/ntcchat.wasm");
export const LEDGER_TYPE = process.env['LEDGER_TYPE'] as "icrc" | "icp";

export {toState};

export async function CanNTCCHAT(pic:PocketIc) {
    const fixture = await pic.setupCanister<NTCCHAT_SERVICE>({
        idlFactory: NTCCHAT_IDL_FACTORY,
        wasm: NTCCHAT_WASM_PATH,
        arg: IDL.encode(ntcchat_init({ IDL }), []),
    });

    return fixture;
}



export {NTCCHAT_SERVICE, NTCCHAT_IDL_FACTORY, ntcchat_init};

export async function CanNTC(pic:PocketIc, targetCanisterId:Principal, ledgerCanisterId:Principal) {
    const fixture = await pic.setupCanister<NTCService>({
        targetCanisterId,
        idlFactory: NTCIdlFactory,
        wasm: NTC_WASM_PATH,
        arg: IDL.encode(ntcInit({ IDL }), [{ledgerId: ledgerCanisterId}]),
    });

    return fixture;
};

export {NTCService, NTCIdlFactory, ntcInit};

export async function CanCycleWallet(pic:PocketIc) {
    const fixture = await pic.setupCanister<CycleWalletService>({
        idlFactory: CycleWalletIdlFactory,
        wasm: CYCLE_WALLET_WASM_PATH,
        arg: IDL.encode(cycleWalletInit({ IDL }), []),
    });

    return fixture;
};

export {CycleWalletService, CycleWalletIdlFactory, cycleWalletInit};


export async function CanBasic(pic:PocketIc, ledgerCanisterId:Principal) {
    
    const fixture = await pic.setupCanister<BasicService>({
        idlFactory: BasicIdlFactory,
        wasm: BASIC_WASM_PATH,
        arg: IDL.encode(basicInit({ IDL }), [{ledgerId: ledgerCanisterId, ledger_type: {[LEDGER_TYPE]:null}}]),
    });

    return fixture;
};
export {BasicService, BasicIdlFactory, basicInit};

export async function CanBurn(pic:PocketIc, ledgerCanisterId:Principal, specific_canister_id:Principal | undefined) {
    
    const fixture = await pic.setupCanister<BurnService>({
        ...specific_canister_id?{targetCanisterId: specific_canister_id}:{},
        idlFactory: BurnIdlFactory,
        wasm: BURN_WASM_PATH,
        arg: IDL.encode(burnInit({ IDL }), [{ledgerId: ledgerCanisterId, ledger_type: {[LEDGER_TYPE]:null}}]),
    });

    return fixture;
};
export {BurnService, BurnIdlFactory, burnInit};

export async function CanFastscan(pic:PocketIc, ledgerCanisterId:Principal) {
    
    const fixture = await pic.setupCanister<FastscanService>({
        idlFactory: FastscanIdlFactory,
        wasm: FASTSCAN_WASM_PATH,
        arg: IDL.encode(fastscanInit({ IDL }), [{ledgerId: ledgerCanisterId, ledger_type: {[LEDGER_TYPE]:null}}]),
    });

    return fixture;
};
export {FastscanService, FastscanIdlFactory, fastscanInit};


export async function CanPassback(pic:PocketIc, ledgerCanisterId:Principal) {
    
    const fixture = await pic.setupCanister<PassbackService>({
        idlFactory: PassbackIdlFactory,
        wasm: PASSBACK_WASM_PATH,
        arg: IDL.encode(passbackInit({ IDL }), [{ledgerId: ledgerCanisterId, ledger_type: {[LEDGER_TYPE]:null}}]),
    });

    return fixture;
};  

export {PassbackService, PassbackIdlFactory, passbackInit};   



let ICRC_WASM_PATH = resolve(__dirname, "./icrc_ledger/ledger.wasm");
if (process.env['LEDGER'] === "motoko") {
    console.log("ðŸš€ðŸ¦€ USING MOTOKO LEDGER - BRACE FOR IMPACT! ðŸ’¥ðŸ¦‘");
    ICRC_WASM_PATH = resolve(__dirname, "./icrc_ledger/motoko_ledger.wasm");
}

let ICP_WASM_PATH = resolve(__dirname, "./icp_ledger/ledger.wasm");

function get_args(me:Principal) {
    let ledger_args:ICRCLedgerArg = {
        Init: {
            minting_account: {
                owner: me,
                subaccount: []
            },
            fee_collector_account: [], //{ owner: me, subaccount:[] }
            transfer_fee: 10000n,
            decimals: [8],
            token_symbol: "tCOIN",
            token_name: "Test Coin",
            metadata: [],
            initial_balances: [], //[{ owner: me, subaccount:[] }, 100000000000n]
            archive_options: {
                num_blocks_to_archive: 1000n,
                trigger_threshold: 3000n,
                controller_id: me,
                max_transactions_per_response: [],
                max_message_size_bytes: [],
                cycles_for_archive_creation: [1000_000_000_000n],
                node_max_memory_size_bytes: [],
            },
            maximum_number_of_accounts: [],
            accounts_overflow_trim_quantity: [],
            max_memo_length: [80],
            feature_flags: [{ icrc2: true }],            
        },
    };


    return ledger_args;
    }



    
export async function ICRCLedger(pic: PocketIc, me:Principal, subnet:Principal | undefined) {

    const fixture = await pic.setupCanister<ICRCLedgerService>({
        idlFactory: ICRCLedgerIdlFactory,
        wasm: ICRC_WASM_PATH,
        arg: IDL.encode(icrcInit({IDL}), [get_args(me)]),
        ...subnet?{targetSubnetId: subnet}:{},
    });

    await pic.addCycles(fixture.canisterId, 100_000_000_000_000_000);   
    
    return {
        canisterId: fixture.canisterId,
        actor: fixture.actor as Actor<ICRCLedgerService>
    };
};


export async function ICRCLedgerUpgrade(pic: PocketIc, me:Principal, canister_id:Principal, subnet:Principal | undefined) {
    await pic.upgradeCanister({ canisterId: canister_id, wasm: ICRC_WASM_PATH, arg: IDL.encode(icrcInit({ IDL }), [{Upgrade: []}]) });

}

export { ICRCLedgerService, ICRCLedgerIdlFactory, icrcInit };

// Helper function to convert ICP Block to ICRC Transaction
function convertICPBlockToICRCTransaction(block: ICPBlock): ICRCTransaction {
    const icpTx = block.transaction;
    
    // Initialize ICRC transaction with common fields
    const icrcTx: ICRCTransaction = {
        burn: [],
        kind: "",
        mint: [],
        approve: [],
        timestamp: block.timestamp.timestamp_nanos,
        transfer: []
    };

    // Convert operation to appropriate ICRC transaction type
    if (icpTx.operation && icpTx.operation.length > 0) {
        const operation = icpTx.operation[0];
        
        if ('Mint' in operation) {
            icrcTx.kind = "mint";
            icrcTx.mint = [{
                to: getAccountFromMap(operation.Mint.to),
                memo: icpTx.icrc1_memo,
                created_at_time: [icpTx.created_at_time.timestamp_nanos],
                amount: BigInt(operation.Mint.amount.e8s)
            }];
        } else if ('Burn' in operation) {
            icrcTx.kind = "burn";
            icrcTx.burn = [{
                from: getAccountFromMap(operation.Burn.from),
                memo: icpTx.icrc1_memo,
                created_at_time: [icpTx.created_at_time.timestamp_nanos],
                amount: BigInt(operation.Burn.amount.e8s),
                spender: operation.Burn.spender ? [getAccountFromMap(operation.Burn.spender[0])] : []
            }];
        } else if ('Transfer' in operation) {
            icrcTx.kind = "transfer";
            icrcTx.transfer = [{
                to: getAccountFromMap(operation.Transfer.to),
                fee: [BigInt(operation.Transfer.fee.e8s)],
                from: getAccountFromMap(operation.Transfer.from),
                memo: icpTx.icrc1_memo,
                created_at_time: [icpTx.created_at_time.timestamp_nanos],
                amount: BigInt(operation.Transfer.amount.e8s),
                spender: operation.Transfer.spender ? [getAccountFromMap(operation.Transfer.spender[0])] : []
            }];
        } else if ('Approve' in operation) {
            icrcTx.kind = "approve";
            icrcTx.approve = [{
                fee: [BigInt(operation.Approve.fee.e8s)],
                from: getAccountFromMap(operation.Approve.from),
                memo: icpTx.icrc1_memo,
                created_at_time: [icpTx.created_at_time.timestamp_nanos],
                amount: BigInt(operation.Approve.allowance.e8s),
                expected_allowance: operation.Approve.expected_allowance ? [BigInt(operation.Approve.expected_allowance[0].e8s)] : [],
                expires_at: operation.Approve.expires_at ? [operation.Approve.expires_at[0].timestamp_nanos] : [],
                spender: getAccountFromMap(operation.Approve.spender)
            }];
        }
    }

    return icrcTx;
}

let registered_accounts = new Map<string, Account>();

export function account2aid(account: Account) {
    return AccountIdentifier.fromPrincipal({
        principal: account.owner,
        subAccount: account.subaccount.length > 0 ? SubAccount.fromBytes(new Uint8Array(account.subaccount[0])) : undefined 
    });
}

// Function to register account mappings - calculates AccountIdentifier from Account
export function registerAccount(account: Account) {
    const accountIdentifier = account.subaccount.length > 0 
        ? AccountIdentifier.fromPrincipal({
            principal: account.owner,
            subAccount: SubAccount.fromBytes(new Uint8Array(account.subaccount[0]))
        })
        : AccountIdentifier.fromPrincipal({
            principal: account.owner
        });
    const key = accountIdentifier.toHex();
    registered_accounts.set(key, account);
}

// Helper function to convert AccountIdentifier to map key
function accountIdentifierToKey(accountIdentifier: Uint8Array | number[] | undefined): string | null {
    if (!accountIdentifier) {
        return null;
    }
    if (accountIdentifier instanceof Uint8Array) {
        return Buffer.from(accountIdentifier).toString('hex');
    } else {
        return Buffer.from(accountIdentifier).toString('hex');
    }
}

// Helper function to get account from map or return default
function getAccountFromMap(accountIdentifier: Uint8Array | number[] | undefined): Account {
    if (!accountIdentifier) {
        return { owner: Principal.anonymous(), subaccount: [] };
    }
    const key = accountIdentifierToKey(accountIdentifier);
    if (!key) {
        return { owner: Principal.anonymous(), subaccount: [] };
    }
    return registered_accounts.get(key) || { owner: Principal.anonymous(), subaccount: [] };
}

export async function realICPLedger(pic: PocketIc, can_id:Principal) {
    const actor = pic.createActor<ICPLedgerService>(ICPLedgerIdlFactory, can_id);
    return actor;
};

export async function ICPLedger(pic: PocketIc, me:Principal, subnet:Principal | undefined) {
    let me_address = AccountIdentifier.fromPrincipal({
        principal: me,
      }).toHex();
      
    let ledger_args:ICPLedgerCanisterPayload = {
        Init: {
            'send_whitelist' : [],
            'token_symbol' : ["tCOIN"],
            'transfer_fee' :[{e8s: 10000n}],
            'minting_account' : me_address,
            'maximum_number_of_accounts' : [],
            'accounts_overflow_trim_quantity' : [],
            'transaction_window' : [],
            'max_message_size_bytes' : [],
            'icrc1_minting_account' : [{owner: me, subaccount: []}],
            'archive_options' : [],
            'initial_values' : [],
            'token_name' : ["Test Coin"],
            'feature_flags' : [],
        },
    };

    const fixture = await pic.setupCanister<ICPLedgerService>({
        idlFactory: ICPLedgerIdlFactory,
        wasm: ICP_WASM_PATH,
        arg: IDL.encode(icpInit({IDL}), [ledger_args]),
        ...subnet?{targetSubnetId: subnet}:{},
    });

    await pic.addCycles(fixture.canisterId, 100_000_000_000_000_000);   

    // Create a proxy that has ICRCLedgerService interface but forwards to ICP ledger
    const proxyActor = new Proxy(fixture.actor, {
        get(target, prop) {
            // If the property exists on the target, return it
            if (prop in target) {
                return target[prop as keyof typeof target];
            }
            
            // Handle custom methods
            if (prop === 'get_transactions') {
                return async function(request: GetTransactionsRequest): Promise<GetTransactionsResponse> {
                    let resp: QueryBlocksResponse = await fixture.actor.query_blocks({start: request.start, length: request.length});

                    // Convert ICP blocks to ICRC transactions
                    const transactions: ICRCTransaction[] = resp.blocks.map((block: ICPBlock) => {
                        return convertICPBlockToICRCTransaction(block);
                    });

                    // Convert archived blocks to archived transactions
                    const archived_transactions = resp.archived_blocks.map(archived => ({
                        start: archived.start,
                        length: archived.length,
                        callback: archived.callback as any // Type conversion for compatibility
                    }));

                    return {
                        first_index: resp.first_block_index,
                        log_length: resp.chain_length,
                        transactions: transactions,
                        archived_transactions: archived_transactions
                    };
                };
            }
            
            // For any other missing method, return undefined or throw an error
            return undefined;
        }
    }) as unknown as Actor<ICRCLedgerService>;
    
    return {
        canisterId: fixture.canisterId,
        actor: proxyActor
    };
};


export { ICPLedgerService, ICPLedgerIdlFactory, icpInit };

export async function Ledger(pic: PocketIc, me:Principal) {
    if (LEDGER_TYPE === "icrc") {
        return await ICRCLedger(pic, me, undefined);
    } else if (LEDGER_TYPE === "icp") {
        return await ICPLedger(pic, me, undefined);
    }
}

export async function LedgerUpgrade(pic: PocketIc, me:Principal, canister_id:Principal) {
    if (LEDGER_TYPE === "icrc") {
        return await ICRCLedgerUpgrade(pic, me, canister_id, undefined);
    } else if (LEDGER_TYPE === "icp") {
        
    }
}

// export type LedgerService = typeof LEDGER_TYPE extends "icrc"
//   ? ICRCLedgerService
//   : ICPLedgerService;

// export type LedgerIdlFactory = typeof LEDGER_TYPE extends "icrc"
//   ? typeof ICRCLedgerIdlFactory
//   : typeof ICPLedgerIdlFactory;


export type LedgerService = ICRCLedgerService;

