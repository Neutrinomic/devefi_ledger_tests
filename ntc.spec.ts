import { Principal } from '@dfinity/principal';

import { Actor, PocketIc, createIdentity } from '@dfinity/pic';
import { IDL } from '@dfinity/candid';

  import { toState, CanBasic, CanNTC, NTCService, ICRCLedger, ICRCLedgerService, Ledger, realICPLedger, ICPLedgerService, CycleWalletService, CanCycleWallet} from './common';


const T = 1_000_000_000_000n;
const NT = T / 10000n;

describe('NTC', () => {
    let pic: PocketIc;
    let ledger: Actor<ICRCLedgerService>; 
    let ledgerCanisterId: Principal;
    let ICPLedger: Actor<ICPLedgerService>;
    let minter: Actor<NTCService>;
    let minterCanisterId: Principal;
    let cycleWallet: Actor<CycleWalletService>;
    let cycleWalletCanisterId: Principal;

    const jo = createIdentity('superSecretAlicePassword');
    const bob = createIdentity('superSecretBobPassword');
  
    beforeAll(async () => {


      pic = await PocketIc.create(process.env.PIC_URL);
      let minter_can_id = Principal.fromText("7ew52-sqaaa-aaaal-qsrda-cai");


      // Ledger
      const ledgerfixture = await ICRCLedger(pic, minter_can_id, undefined);
      ledger = ledgerfixture.actor;
      ledgerCanisterId = ledgerfixture.canisterId;
      
      const fixture1 = await CanNTC(pic, minter_can_id, ledgerCanisterId);
      minter = fixture1.actor;
      minterCanisterId = fixture1.canisterId;

      const cycleWalletfixture = await CanCycleWallet(pic);
      cycleWallet = cycleWalletfixture.actor;
      cycleWalletCanisterId = cycleWalletfixture.canisterId;

      
      // ledger.setPrincipal(jo.getPrincipal());
      await passTime(30);

    });
  
    afterAll(async () => {
      await pic.tearDown();
    });

    it(`Initial checks`, async () => {
      expect(minterCanisterId.toText()).toBe("7ew52-sqaaa-aaaal-qsrda-cai");

      let symbol = await ledger.icrc1_symbol();
      let decimals = await ledger.icrc1_decimals();
      let minter = await ledger.icrc1_minting_account();
      let name = await ledger.icrc1_name();
      let fee = await ledger.icrc1_fee();

      expect(decimals).toBe(8);
      expect(minter[0].owner.toText()).toBe(minterCanisterId.toText());
      expect(fee).toBe(10000n);
      
    });
    it(`Adding, burning cycles in cycle wallet`, async () => {
      const cyclesBalance = await cycleWallet.getCyclesBalance();
      expect(cyclesBalance).toBeGreaterThan(999999n*T);

      //add cycles
      const newCyclesBalance = await pic.addCycles(cycleWalletCanisterId, Number(100n*T));

      const cyclesBalance2 = await cycleWallet.getCyclesBalance();
      expect(cyclesBalance2).toBeGreaterThan(999999n*T + 100n*T);
      expect(cyclesBalance2).toBeLessThan(999999n*T + 101n*T);

      await cycleWallet.burn_my_cycles(ledgerCanisterId);

      const cyclesBalance3 = await cycleWallet.getCyclesBalance();
      expect(cyclesBalance3).toBeLessThan(T / 2n);

      await pic.addCycles(cycleWalletCanisterId, Number(1000n*T));
      const cyclesBalance4 = await cycleWallet.getCyclesBalance();
      expect(cyclesBalance4).toBeGreaterThan(1000n*T);


    });

    it(`Get minter stats`, async () => {
      const stats = await minter.stats();
      expect(stats.failed_topups).toBe(0n); 
      expect(stats.cycles).toBeGreaterThan(999999n*T);
      expect(stats.cycles).toBeLessThan(1000000n*T);
      expect(stats.topped_up).toBe(0n);
      
    });
  
    it(`Mint NTC`, async () => {
      await cycleWallet.mint_ntc(1n*T, {owner:bob.getPrincipal(), subaccount:[]});

      expect(await cycleWallet.getCyclesBalance()).toBeLessThan(1000n*T);

      await passTime(20);

      let log = await ledger.get_transactions({start:0n, length:100n});
      expect(log.log_length).toBe(1n);

      expect(toState(log.transactions[0])).toMatchSnapshot();

      // Check ledger balance 
      const ledgerBalance = await ledger.icrc1_balance_of({owner:bob.getPrincipal(), subaccount:[]});
      expect(ledgerBalance).toBe(99990000n);


    }, 600*1000);

    it(`Get minter stats`, async () => {
      const stats = await minter.stats();
      expect(stats.failed_topups).toBe(0n); 
      expect(stats.cycles).toBeGreaterThan(1000000n*T);
      expect(stats.topped_up).toBe(0n);
      
    });
    
    it(`Mint lots of NTC`, async () => {

      for (let i=0; i<100; i++) {
        await cycleWallet.mint_ntc(1n*T, {owner:bob.getPrincipal(), subaccount:[]});
      }
     
      await passTime(10);
      let log = await ledger.get_transactions({start:0n, length:101n});
      expect(log.log_length).toBe(101n);

    }, 600*1000);

    it(`Check ledger balance`, async () => {
      const ledgerBalance = await ledger.icrc1_balance_of({owner:bob.getPrincipal(), subaccount:[]});
      expect(ledgerBalance).toBe(10098990000n);
    });

it(`Burn NTC`, async () => {


  let wallet_balance = await cycleWallet.getCyclesBalance();

  expect(wallet_balance/T).toBe(899n);

  let topup_account_resp = await minter.get_account( cycleWalletCanisterId);
  let topup_account = topup_account_resp[0];
  let topup_account_text = topup_account_resp[1];

  expect(topup_account_resp).toMatchSnapshot();
  
  // Bob sends NTC to topup account
  ledger.setPrincipal(bob.getPrincipal());
  let resp = await ledger.icrc1_transfer({
    to: topup_account,
    from_subaccount: [],
    amount: 10n * NT,
    fee: [],
    memo: [],
    created_at_time: [],
  });

  expect(toState(resp).Ok).toBe("101");

  await passTime(5);

  let wallet_balance_after = await cycleWallet.getCyclesBalance();

  expect(wallet_balance_after/T).toBe((wallet_balance + 10n*T - T/10n)/T);

  expect(wallet_balance_after/T).toBe(899n + 9n);
});


it(`Make a lot of burn requests`, async () => {

  await cycleWallet.mint_ntc(100n*T, {owner:bob.getPrincipal(), subaccount:[]});
  await passTime(2);
  
  let wallet_balance = await cycleWallet.getCyclesBalance();

  expect(wallet_balance/T).toBe(808n);

  let topup_account_resp = await minter.get_account( cycleWalletCanisterId);
  let topup_account = topup_account_resp[0];
  let topup_account_text = topup_account_resp[1];

  expect(topup_account_resp).toMatchSnapshot();
  
  // Bob sends NTC to topup account
  ledger.setPrincipal(bob.getPrincipal());
  for (let i=0; i<100; i++) {
    let resp = await ledger.icrc1_transfer({
      to: topup_account,
      from_subaccount: [],
      amount: 1n * NT,
      fee: [],
      memo: [],
      created_at_time: [],
    });

    expect(toState(resp).Err).not.toBeDefined();
  };

  
  await passTime(40);

  let wallet_balance_after = await cycleWallet.getCyclesBalance();
  let one_fill = 1n*T - T/10n;
  expect(wallet_balance_after/T).toBe((wallet_balance + one_fill*100n)/T);
expect((wallet_balance + one_fill*100n)/T).toBe(808n + 90n);

});

it(`Check stats for burned and failed topups`, async () => {

  let stats = await minter.stats();
  expect(stats.failed_topups).toBe(0n);
  expect(stats.cycles/T).toBeGreaterThan(1000000n);
  expect(stats.topped_up/T).toBe(99n);
  

});


it(`Check if minter queue is empty and all is sent`, async () => {
  let queue = await minter.get_queue();
  expect(queue.length).toBe(0);
});


it(`Send to non existing canister`, async () => {
  let topup_account_resp = await minter.get_account(Principal.fromText("togwv-zqaaa-aaaal-qr7aa-cai"));
  let topup_account = topup_account_resp[0];
  let topup_account_text = topup_account_resp[1];

  let resp = await ledger.icrc1_transfer({
    to: topup_account,
    from_subaccount: [],
    amount: 10n * NT,
    fee: [],
    memo: [],
    created_at_time: [],
  });

  expect(toState(resp).Err).not.toBeDefined();
  await passTime(10);
  let queue = await minter.get_queue();
  expect(queue.length).toBe(1);

});

it(`Wait for it to fail`, async () => {
  
  let stats_before = await minter.stats();
  expect(stats_before.topped_up/T).toBe(99n);
  expect(stats_before.cycles/T).toBe(1000101n);
  expect(stats_before.failed_topups).toBe(0n);
  for (let i=0; i<100; i++) {
    await pic.advanceTime(60*1000);
    await pic.tick(1);
  }
  
  let queue = await minter.get_queue();
  expect(queue.length).toBe(0);

  let stats = await minter.stats();
  expect(stats.topped_up/T).toBe(99n);
  expect(stats.cycles/T).toBe(1000101n);
  expect(stats.failed_topups/T).toBe(9n);

});


  async function passTime(n:number) {
    for (let i=0; i<n; i++) {
      await pic.advanceTime(3*1000);
      await pic.tick(2);
    }
  }

});
