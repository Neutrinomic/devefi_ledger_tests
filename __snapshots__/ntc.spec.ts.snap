// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NTC Burn NTC 1`] = `
[
  {
    "owner": {
      "__principal__": "7ew52-sqaaa-aaaal-qsrda-cai",
    },
    "subaccount": [
      Uint8Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        255,
        255,
        255,
        255,
        255,
        224,
        0,
        1,
        1,
        1,
        10,
      ],
    ],
  },
  "7ew52-sqaaa-aaaal-qsrda-cai-jb4bgmy.ffffffffffe0000101010a",
  {
    "__principal__": "lqy7q-dh777-77777-aaaaq-cai",
  },
]
`;

exports[`NTC Make a lot of burn requests 1`] = `
[
  {
    "owner": {
      "__principal__": "7ew52-sqaaa-aaaal-qsrda-cai",
    },
    "subaccount": [
      Uint8Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        255,
        255,
        255,
        255,
        255,
        224,
        0,
        1,
        1,
        1,
        10,
      ],
    ],
  },
  "7ew52-sqaaa-aaaal-qsrda-cai-jb4bgmy.ffffffffffe0000101010a",
  {
    "__principal__": "lqy7q-dh777-77777-aaaaq-cai",
  },
]
`;

exports[`NTC Mint NTC 1`] = `
{
  "approve": [],
  "burn": [],
  "kind": "mint",
  "mint": [
    {
      "amount": "99990000",
      "created_at_time": [
        "1620328720000000100",
      ],
      "memo": [
        "000000000000000000000000000000000000000000ffffffffffe0000101010a",
      ],
      "to": {
        "owner": "ydl4r-asr5o-7axs3-tshas-4xugy-bvg4x-ixnjd-6qex3-guw6d-5pahc-oqe",
        "subaccount": [],
      },
    },
  ],
  "timestamp": "1620328723000000300",
  "transfer": [],
}
`;
